domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java

```diff
@@ -1739,57 +1735,41 @@ QueryHelper buildQueryHelper(
       addIdField = "";
     }
 
-    QueryHelper queryHelper = new QueryHelper(transactionMode, table, facets);
+    QueryHelper queryHelper = new QueryHelper(transactionMode, table);
 
     String countOn = "*";
     String distinctOnClause = "";
     if (distinctOn != null && !distinctOn.isEmpty()) {
       distinctOnClause = String.format("DISTINCT ON(%s) ", distinctOn);
       countOn = String.format("DISTINCT(%s)", distinctOn);
     }
-
     queryHelper.selectQuery = SELECT + distinctOnClause + fieldName + addIdField
       + FROM + schemaName + DOT + table + SPACE + wrapper.toString();
-    queryHelper.countQuery = SELECT + "COUNT(" + countOn + ") "
+    queryHelper.countQuery = SELECT + "COUNT(" + countOn + ")"
       + FROM + schemaName + DOT + table + SPACE + wrapper.getWhereClause();
+    String mainQuery = SELECT + distinctOnClause + fieldName + addIdField
+      + FROM + schemaName + DOT + table + SPACE + wrapper.getWithoutLimOff();
 
-    ParsedQuery pq = new ParsedQuery();
-    pq.setCountQuery(queryHelper.countQuery);
-    if (!wrapper.getWhereClause().isEmpty()) {
-      pq.setWhereClause(wrapper.getWhereClause());
-    }
-    if (!wrapper.getLimit().toString().isEmpty()) {
-      pq.setLimitClause(wrapper.getLimit().toString());
-    }
-    if (!wrapper.getOffset().toString().isEmpty()) {
-      pq.setOffsetClause(wrapper.getOffset().toString());
-    }
-    pq.setQueryWithoutLimOff(SELECT + distinctOnClause + fieldName + addIdField
-      + FROM + schemaName + DOT + table + SPACE + wrapper.getWithoutLimOff());
-
-    String offsetClause = null;
-    if (queryHelper.facets != null && !queryHelper.facets.isEmpty() && queryHelper.table != null) {
-      FacetManager facetManager = buildFacetManager(queryHelper.table, pq, queryHelper.facets);
+    if (facets != null && !facets.isEmpty() && queryHelper.table != null) {
+      FacetManager facetManager = buildFacetManager(wrapper, queryHelper, mainQuery, facets);
       // this method call invokes freemarker templating
       queryHelper.selectQuery = facetManager.generateFacetQuery();
-      queryHelper.countQuery = facetManager.getCountQuery();
-
-      offsetClause = facetManager.getOffsetClause();
-    } else {
-      offsetClause = pq.getOffsetClause();
     }
-    if (offsetClause != null) {
-      Matcher matcher = OFFSET_MATCH_PATTERN.matcher(offsetClause);
-      if (matcher.find()) {
-        queryHelper.offset = Integer.parseInt(matcher.group(1));
-      }
+    if (distinctOnClause.equals("*")) { // only estimate on non distinct cases
+      queryHelper.countQuery = SELECT + "count_estimate_smart('"
+        + org.apache.commons.lang.StringEscapeUtils.escapeSql(queryHelper.countQuery)
+        + "')";
```

Comment:
These three lines have no test coverage. Can you write an integration test that runs this with same sample data in the database or can you at least write a unit test for this method?

Reply:
Yes.. My bad.. Of course these lines should be in use..